### PR TITLE
Fix Double Application of Softmax for Router Logits in MoE models

### PR DESCRIPTION
**Pull request by ionut-anghelina**

### CODE REVIEW

File: tests/models/mixtral/test_modeling_mixtral.py [hunk 1]
```diff
@@ -89,6 +89,14 @@ def test_load_balancing_loss(self):
         self.assertEqual(result.router_logits[0].shape, (91, config.num_local_experts))
         torch.testing.assert_close(result.aux_loss.cpu(), torch.tensor(2, dtype=torch.float32), rtol=1e-2, atol=1e-2)
 
+        # Verify router_logits are raw logits, not softmax probabilities (regression test for double-softmax bug)
```
Comment (vasqu) [inline tests/models/mixtral/test_modeling_mixtral.py:92]: Iirc, we have more appearances of that test in other models. It doesnt hurt to add them to all we have + maybe make it a generalized one in causal lm tester (because we now have ways to properly detect moes with the interface)

**Comment by vasqu** [comment]
@Rocketknight1 I'm not sure about the current state here so just left a comment here since it seemed the most recent state of things. Lmk if not or where I should properly look at

**Comment by vasqu** [comment]
Let's also add the fixes and closes statements for the issue and other PR please

**Comment by vasqu** [comment]
Ok #45131 was merged instead, we can still use this PR for tests tho @Rocketknight1 @ionut-anghelina

**Comment by ErenAta16** [comment]
This diff only adds tests, `test_modeling_mixtral.py` and `test_modeling_qwen2_moe.py`, no source change in either model's router code.

Checked both `MixtralTopKRouter.forward` and `Qwen2MoeTopKRouter.forward` against current `main`: both compute `router_logits = F.linear(hidden_states, self.weight)` and return that raw value as the first element (the one captured by `OutputRecorder(..., index=0)` into `outputs.router_logits`). `router_probs = softmax(router_logits, ...)` is a separate local variable used only to compute top-k routing weights, it's never what gets returned as `router_logits`. So on current `main`, `router_logits` already looks like raw logits, not softmax output, which is what the new regression tests assert.

Where's the actual double-softmax happening? If it's in `load_balancing_loss_func` (e.g. if `router_logits` ever gets passed in already-softmaxed from some other call site) rather than in the router forward itself, that source fix seems to be missing from this diff, only the router-output-level regression test is here. Want to make sure I'm not missing the actual repro before assuming this needs a source change added.

**Comment by vasqu** [comment]
#45131 fixed it, and then this PR kind of got abandoned because it aimed to do the same :sweat_smile:

**Comment by ErenAta16** [comment]
That explains it, #45131 (merged 2026-04-13) is exactly what I was seeing on current `main`: both routers already return raw logits, not probabilities. Confirms the analysis rather than contradicts it, the fix landed, this PR's tests just never got reconciled with that. Makes sense to close this one as superseded.
